### PR TITLE
Datapath portion of egress label-based policy enforcement

### LINKED ISSUE #2553: Label-based egress support: datapath implementation
Parent: #1488 

Augment the BPF datapath to be able to handle label-based egress. This will consist of reading from a table that is populated by the Cilium agent with pod IPs to which egress traffic is allowed. 

* [x] Add an IP->SecID mapping table to BPF (#2954)
* [x] Add Label-based lookup to the map, following the example of ingress policy map (#2954)
* [x] Refactor the existing L4 port policy from code to a native BPF map (L4-only)  (#2918)
* Extra wishlist items, but not required to implement:
  * [ ] Userspace - populate maps with `proxy_port` (#2564)
  * [ ] Userspace - atomic map replace

### PR DESCRIPTION
**Pull request by joestringer**
Fixes: #2553

This PR implements egress label-based policies (labels + labels-dependent l4) in the datapath. It's prepared in a way that allows it to be merged independent of the rest of the label-based egress policy work. The last two commits should be fixed up when label-based egress support lands in the tree.

The core change is that one of the pad bits in `struct policy_key` now represents whether the entry represents ingress (0) or egress (1). Because the entire field is currently padding, and is set to 0, this allows the change to be backward-compatible - earlier versions will only set the bit to 0, indicating ingress, and upon upgrade the entries will be treated exactly the same.

In addition to this change, there is a new map, `cilium_remote_lxc`, which maps destination IP addresses to remote identites. This will be populated by subsequent changes in the userspace.

Note that when the `egress` bit is set in the `policy_key` for the `POLICY_MAP`, the `sec_label` represents the destination identity rather than the source identity.

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by aanm** [comment]
@joestringer is there a duplicate commit from https://github.com/cilium/cilium/pull/2992? It seems some of the code is similar

**Comment by joestringer** [comment]
@aanm yes, one commit is the same. This work requires the l4-only bug to be fixed or current l4 egress policy breaks.

**Comment by joestringer** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

File: bpf/lib/maps.h [hunk 1]
```diff
@@ -310,6 +310,17 @@ static __always_inline int lpm4_egress_lookup(__be32 addr)
...
 #endif
 
+#if defined POLICY_EGRESS && defined LXC_ID
+/* Global IP -> Identity map for applying egress label-based policy */
+struct bpf_elf_map __section_maps cilium_remote_lxc = {
+	.type		= BPF_MAP_TYPE_HASH,
+	.size_key	= sizeof(struct endpoint_key),
+	.size_value	= sizeof(struct remote_endpoint_info),
+	.pinning	= PIN_GLOBAL_NS,
+	.max_elem	= ENDPOINTS_MAP_SIZE, /* XXX: Consider resizing? */
```
Comment (tgraf) [inline bpf/lib/maps.h:320]: I think it's fine but bumping it to 250K doesn't hurt either. The map is still quite small. Please create a GH to track adjusting this on demand. Given regenerations, we could easily resize the map.

**Review by tgraf — APPROVED** [review]
Awesome work

File: bpf/lib/maps.h (hunk 1 above)
Comment (joestringer) [reply bpf/lib/maps.h:320]: Created #3075. Main reason I put this comment was that it's now referring to remote containers so binding it to the same limit as the lxcmap might not be appropriate.

**Pull request merged**